### PR TITLE
84 localizable message strings

### PR DESCRIPTION
**Pull request by EasyRhin0**
Wanted to point out the change in ResultLogJsonWriter.WriteRules. Not sure if what I did is OK,
(Sarif\Writers\ResultLogJsonWriter.cs)

### CODE REVIEW

File: src/Sarif.Driver.UnitTests/Sdk/FormatForVisualStudioTests.cs [hunk 1]
```diff
@@ -15,18 +15,18 @@ namespace Microsoft.CodeAnalysis.Sarif.Driver
     public class FormatForVisualStudioTests
     {
         private const string TestRuleId = "TST0001";
-        private const string TestTemplateId = "testTemplateId";
+        private const string TestMessageStringId = "testMessageStringId";
```
Comment (lgolding) [inline src/Sarif.Driver.UnitTests/Sdk/FormatForVisualStudioTests.cs]: Hah, nice rename :-)

File: src/Sarif.Driver/Sdk/SkimmerBase.cs [hunk 2]
```diff
@@ -16,64 +16,62 @@ public SkimmerBase()
...
-        virtual public IDictionary<string, string> MessageTemplates
+        virtual public IDictionary<string, string> MessageStrings
         {
             get
             {
-                if (this.messageTemplates == null)
+                if (this.messageStrings == null)
                 {
-                    this.messageTemplates = InitializeMessageTemplates();
+                    this.messageStrings = InitializeMessageTemplates();
```
Comment (lgolding) [inline src/Sarif.Driver/Sdk/SkimmerBase.cs]: `InitializeMessageTemplates` => `InitializeMessageStrings`.

Search "template" case-insensitive throughout the code base to catch these stragglers.

File: src/Sarif.Driver/Sdk/SkimmerBase.cs [hunk 3]
```diff
@@ -16,64 +16,62 @@ public SkimmerBase()
...
-        virtual public IDictionary<string, string> RichMessageTemplates
+        virtual public IDictionary<string, string> RichMessageStrings
         {
             get
             {
-                if (this.richMessageTemplates == null)
+                if (this.richMessageStrings == null)
                 {
-                    this.richMessageTemplates = InitializeRichMessageTemplates();
+                    this.richMessageStrings = InitializeRichMessageTemplates();
```
Comment (lgolding) [inline src/Sarif.Driver/Sdk/SkimmerBase.cs]: `InitializeRichMessageStrings`

File: src/Sarif.FunctionalTests/ConverterTestData/FxCop/FxCopReportAllMessageStates.xml.sarif [hunk 4]
```diff
@@ -131,6 +136,7 @@
         },
         {
           "ruleId": "CA1824",
+          "message": {},
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/ConverterTestData/FxCop/FxCopReportAllMessageStates.xml.sarif]: Why emitting empty `message` object?

File: src/Sarif.FunctionalTests/SpecExamples/RuleMetadata.sarif [hunk 5]
```diff
@@ -1,25 +1,55 @@
...
-          "id": "BA2006",
-          "name": "BuildWithSecureTools",
-          "shortDescription": "Application code should be compiled with the most up-to-date tool sets.",
-          "fullDescription": "Application code should be compiled with the most up-to-date tool sets. The latest version is 2.2.",
-          "messageTemplates": {
-            "Error_BadModule": "built with {0} compiler version {1} (Front end version {2})",
-            "Pass": "{0} was built with tools that satisfy configured policy.",
-            "Error": "{0} was compiled with one or tools that do not satisfy configured policy.",
-            "NotApplicable_InvalidMetadata": "{0} was not evaluated for check '{1}'."
+      "results": [
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/SpecExamples/RuleMetadata.sarif]: There should not be a `results` property in a run whose purpose is solely to carry rule metadata.

File: src/Sarif.FunctionalTests/SpecExamples/RuleMetadata.sarif [hunk 6]
```diff
@@ -1,25 +1,55 @@
...
+                "region": {
+                  "startLine": 15
+                }
+              },
+              "fullyQualifiedLogicalName": "collections::list:add"
+            }
+          ]
+        }
+      ],
+      "resources": {
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/SpecExamples/RuleMetadata.sarif]: Only the `resources` property should be present in the `run` object.

File: src/Sarif/CodeGenHints.json [hunk 7]
```diff
@@ -213,6 +213,28 @@
       }
     }
   ],
+  "resources": [
```
Comment (lgolding) [inline src/Sarif/CodeGenHints.json]: Nice!

File: src/Sarif/Schemata/Sarif.schema.json [hunk 8]
```diff
@@ -402,12 +392,7 @@
 
         "description": {
           "description": "A plain text message that describes the proposed fix, enabling viewers to present the proposed change to an end user.",
```
Comment (lgolding) [inline src/Sarif/Schemata/Sarif.schema.json]: Remove the words "plain text".

File: src/Sarif/Schemata/Sarif.schema.json [hunk 9]
```diff
@@ -1260,7 +1239,7 @@
           }
         }
       },
-      "required": [ "tool" ]
+      "required": [ "results", "tool" ]
```
Comment (lgolding) [inline src/Sarif/Schemata/Sarif.schema.json]: Sorry!!! Let's remove `result` from this list. My bad. I need to talk more to Michael about this.

File: src/Sarif/Writers/ResultLogJsonWriter.cs [hunk 10]
```diff
@@ -173,9 +173,14 @@ public void WriteRules(IDictionary<string, IRule> rules)
             EnsureResultsArrayIsNotOpen();
             EnsureStateNotAlreadySet(Conditions.Disposed | Conditions.RulesWritten);
 
+            _jsonWriter.WritePropertyName("resources");
```
Comment (lgolding) [inline src/Sarif/Writers/ResultLogJsonWriter.cs]: Nice!

File: src/Sarif/Writers/SarifLogger.cs [hunk 11]
```diff
@@ -253,10 +253,7 @@ public void Log(IRule rule, Result result)
                 return;
             }
 
-            if (rule != null)
-            {
-                Rules[result.RuleKey ?? result.RuleId] = rule;
-            }
+            Rules[result.RuleKey ?? result.RuleId] = rule;
```
Comment (lgolding) [inline src/Sarif/Writers/SarifLogger.cs]: Why did you remove the `null` test?

File: src/Sarif.Driver/Sdk/SkimmerBase.cs (hunk 2 above)
Comment (EasyRhin0) [reply src/Sarif.Driver/Sdk/SkimmerBase.cs]: Yah meant to do this...

Comment (EasyRhin0) [reply src/Sarif.Driver/Sdk/SkimmerBase.cs]: And done.

File: src/Sarif.Driver/Sdk/SkimmerBase.cs (hunk 3 above)
Comment (EasyRhin0) [reply src/Sarif.Driver/Sdk/SkimmerBase.cs]: Done

File: src/Sarif.FunctionalTests/ConverterTestData/FxCop/FxCopReportAllMessageStates.xml.sarif (hunk 4 above)
Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/ConverterTestData/FxCop/FxCopReportAllMessageStates.xml.sarif]: No idea! Gone.

Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/ConverterTestData/FxCop/FxCopReportAllMessageStates.xml.sarif]: Actually this was because of a null result.message. Added a conditional in the FxCop converter.
